Allow to synchronize groups via OpenID Connect

NobodysNightmare opened the pull request:
When the corresponding claim is present and synchronization of groups is enabled, then we'll import groups from the groups claim.

There is no UI yet to configure group synchronization and it's disabled by default

# Ticket
https://community.openproject.org/wp/58408

# Merge checklist

- [x] Added/updated tests
- [x] Tested manually

NobodysNightmare committed:
Allow to synchronize groups via OpenID Connect

When the corresponding claim is present and synchronization of groups
is enabled, then we'll import groups from the groups claim.

NobodysNightmare commented on modules/openid_connect/spec/services/openid_connect/groups/sync_service_spec.rb:
```diff
@@ -0,0 +1,239 @@
+    it "links the OIDC group to the local group with correct names on both ends" do
+      call_service
+
+      expect(OpenIDConnect::GroupLink.find_by(oidc_name: "/abc/def/ghi")&.group&.name).to eq("de")
```
To be clarified: Should the `oidc_name` be the projected name (i.e. `de`) or should it be the full original group name as present in the claim?

NobodysNightmare replied:
Clarified: Current customer would expect the projected name. And it kinda makes sense, because otherwise we'd build the projection logic only for something that you could do through manual renaming anyways...

ba1ash commented on modules/openid_connect/app/services/openid_connect/groups/sync_service.rb:
```diff
@@ -0,0 +1,111 @@
+      end
+
+      def authentication_provider
+        @user.authentication_provider # TODO: should we find that differently? (risk of multiple auth providers)
```
It should be fine to get auth_provider like that. This service is called after user_auth_provider_link has been updated already. So, it should return "right" auth_provider.
But maybe it makes sense to consider rely on provider which is found explicitly by slug from auth_hash.

NobodysNightmare replied:
True, I forgot about our updating of the link. This means even if there are multiple, we'd luckily get the correct one.

Essentially during implementation of this, I noticed that in other places (like token exchange) we are pretty optimistic that a user will have exactly one auth provider. If this assumption ever becomes wrong, we risk working on the wrong auth provider in different places of the code. What would be better is if we could identify the provider for the current session. But I guess that's an improvement for "later".

ba1ash commented on modules/openid_connect/app/services/openid_connect/groups/sync_service.rb:
```diff
@@ -0,0 +1,111 @@
+        group
+      end
+
+      def remove_groups_except(keep_groups)
```
Here we do not care about cleaning up `oidc_group_memberships` instances, because it is gone alongside with`group_user` instance, right?

NobodysNightmare replied:
Exactly. Deletion of the group user will cascade and take the membership with it.

However, you made me realize that this case wasn't tested yet and I added a test for it.

ba1ash commented on modules/openid_connect/db/migrate/20250710133700_add_oidc_group_links.rb:
```diff
@@ -0,0 +1,43 @@
+      t.belongs_to :group, null: false, index: true, foreign_key: { to_table: :users, on_delete: :cascade }
+
+      t.belongs_to :auth_provider, null: false, index: false, foreign_key: { on_delete: :cascade }
+      t.string :oidc_name, null: false
```
I suggest the following renaming. `oidc_name`  seems to be not specific enough. It can be confused with OIDC provider name, can't it?
WDYT?
```suggestion
      t.string :oidc_group_name, null: false
```

NobodysNightmare replied:
That confusion didn't occur to me until you brought it up :sweat_smile: 

I have no strong objections against the renaming, except that it is longer than the original and makes lines longer and will probably result in new style violations afterwards xD

Did you find the name confusing yourself, when you encountered it, or was that remark out of a matter of caution? Maybe in other words: Would you rate it as a :green_circle: or a :yellow_circle: ?

ba1ash replied:
The name confused me at first and made me think what it is. And then I relized most likely it means group name on OIDC side. So, I rate it as   🟡

NobodysNightmare replied:
changed to `oidc_group_name`

ba1ash commented on modules/openid_connect/spec/models/provider_spec.rb:
```diff
@@ -54,4 +54,51 @@
+      let(:group_prefixes) { ["a"] }
+      let(:group_regexes) { [/[b]/] }
+
+      it { is_expected.to eq([/^a(.+)$/]) }
```
Let's be explicit here and have something like:
```suggestion
      it "prefers prefixes over regular expressions" do
        is_expected.to eq([/^a(.+)$/])
      end
```

ba1ash commented on modules/openid_connect/spec/services/openid_connect/groups/sync_service_spec.rb:
```diff
@@ -0,0 +1,239 @@
+
+      it { is_expected.to be_success }
+
+      it "doesn't mark the connection between user and group to be OIDC-related twice" do
```
Do we need this test case? The database forbids double linking right, right? It seems to be enough to have above the above test case for service call to be successful.

NobodysNightmare replied:
I'd say that the database forbids this is part of the test scope. E.g. I could have forgotten to make the corresponding index unique. Or during the next refactoring, someone adds another column to the unique index, and it is not as unique as it needs to be anymore.

ba1ash commented on modules/openid_connect/app/services/openid_connect/groups/sync_service.rb:
```diff
@@ -0,0 +1,111 @@
+        @result = ServiceResult.success
+      end
+
+      def call(groups_claim:)
```
I have a feeling that this service `call`(or `process_oidc_group` method, not sure yet) have to be transactional.
Let's consider the following scenario: in the middle of `process_oidc_group` group has been deleted by admin. I am not sure there will be no error for the user trying to log it in this case.
Have you considered cases like that already?

NobodysNightmare replied:
That's a good point and probably valid. Though I have to admit that my biggest fear w.r.t. consistency so far is the way that changes to group memberships happen:

```ruby
current_user_ids = group.group_users.pluck(:user_id)
new_user_ids = yield current_user_ids
@result.merge!(
  ::Groups::UpdateService.new(user: User.system, model: group).call(user_ids: new_user_ids)
)
```

What we effectively want to do is adding or removing a single user from a group. However, what we do instead is updating the group with a list of all its users. This leads to (yet unaddressed) race conditions. For example:

* User B is currently in group G, A is not
* B logs in and we learn it must be removed,
* A logs in and we learn it  must be added
* B fetches list of current users
* A fetches list of current users
* B removes itself and saves new group list
* A adds itself and saves new group list

In this execution order, we have A and B in the group after, even though only A should be in the group.

Until now I thought that I'd have to write a new service to make this scenario safe. But what's probably enough would be to add proper locking mechanisms inside the `update_group_users` method.

NobodysNightmare replied:
I mostly addressed the case outlined by me above using row-level locking now.

> Let's consider the following scenario: in the middle of process_oidc_group group has been deleted by admin. I am not sure there will be no error for the user trying to log it in this case.

If this happens at the wrong time, an error could occur, I agree. (time being: after we found that the group exists, but before we try to add the user to it)
I am not yet sure whether this is an edge case we should consider specifically. I'd already imagine it a rare case that an admin deletes a group at all. But even more so, I'd find it surprising for an admin to delete a group that's still being provisioned through OIDC.

ba1ash replied:
Yeah, it is a rare case. And it should be resolved on the next user login. I brought it up just as an example to raise awareness. So, we can leave it be in my opinion.

NobodysNightmare committed:
Add row-level locking for group updates

This should effectively ensure that only one transaction
at a time will update group memberships, and thus should
avoid race conditions where a user removal is accidentially
reverted by another transaction running in parallel.

Caveat: Since the locking is based on row-level locks, it's
ineffective for empty groups. So if two users are added to the same
group in parallel as the very first user, conflicts might still arise.
I consider this a rare enough edge case for now.
